working/terminating-tokens/feature-specification.md

```diff
@@ -0,0 +1,1665 @@
+
+```dart
+a
+-b
+```
+
+That could be parsed as either:
+
+```dart
+a;
+-b;
+```
+
+or:
+
+```dart
+a - b;
+```
+
+
+
+In all cases, we follow the eagerness principle and treat a newline as
+significant before these operators. That means one of these operators at the
+beginning of a line is treated like a unary operator if possible.
+
+Note that these are *only* a problem at the "top level" of a statement
+expression. Once the operator is nested inside parentheses or an argument list
+or something, a semicolon is never allowed and we know we can ignore the
+newline.
+
+The set of operators affected are:
+
+*   `-`, which can be used for both unary negation and infix subtraction:
+
+    ```dart
+    -negate
+    sub - tract
+    ```
+
+    There are **10,551** infix `-` operators in the corpus and **13** (0.123%)
+    are at the beginning of a line. Those thirteen are all in an expression
+    context where we know the newline cannot be significant, so are OK too.
+
+*   `<`, which can be used for generic collection literals and comparison:
+
+    ```dart
+    <int>[]
+    less < than
+    ```
+
+    There are **26,049** infix `<` operators and none at the beginning of a
+    line.
+
+*   `(`, which can be a parenthesized expression or a function invocation:
+
+    ```dart
+    (parenthesized)
+    function(call)
+    ```
+
+    There are **1,437,185** invocations in the corpus, none of which have the
+    `(` on the next line.
+
+*   `[`, which can be a list literal or an index operator:
+
+    ```dart
+    [list, literal]
+    list[access]
+    ```
+
+    There are **173,429** index operator calls in the corpus, **2** (0.001%) of
+    which have the `[` on the next line. Both are the result of chained index
+    operators like:
+
+    ```dart
+    exifData[tag] = ExifConstants.stringValues["Components"]
+            [exifData[tag][0]]
+    ```
+
+    One is in an expression context where the newline is ignored and one is not.
+
+#### Adjacent strings
+
+Adjacent strings are an odd corner of the language. They were added during a
+period of time when Dart disallowed `+` on strings. Now that it supports that,
+there is little reason to also support adjacent strings, but the language still
+does and using it is idiomatic.
+
+This leads to obvious ambiguity. This:
+
+```dart
+"one"
+"two"
+```
+
+Could be:
+
+```dart
+"one";
+"two";
+```
+
+or:
+
+```dart
+"one" "two";
+```
+
+This is the one case where eagerness is obviously wrong. The former
+interpretation can't accomplish anything useful. And, in fact, newlines are the
+primary use case for splitting a string literal into multiple adjacent strings:
+
+*   The corpus has **11,256** adjacent strings. By that, I mean strings that
+    immediately follow another string literal, so `"a" "b" "c"` is two adjacent
+    strings.
+
+*   **11,222** (99.698%) have a newline between them and the previous string.
+
+So, in this case, we don't follow eagerness and ignore a newline between
+adjacent strings.
```

Comment:
So we will mis-interpret:
```dart
var x = "some string"
"some other string".charCodes.forEach(importantAction)
```
?
What can the user do to avoid this? Parenthesizing either string literal. Use a literal semicolon (that's the smaller solution, so that's what I would do, but then the formatter cannot remove it).
That's also a thing to consider: There will be semicolons that cannot be removed (empty statements can be turned into `{}`, but not removed). If we remove the empty statement  `;` completely, and reserve only `{}` for that (the *other* empty statement), then that issue should be gone.
(Let's do that no matter what! Please!)

Reply:
Oof, yes. The proposal will elide the newline there.

> What can the user do to avoid this? Parenthesizing either string literal.

Yeah, parenthesizing would work.

> (Let's do that no matter what! Please!)

SGTM. Empty statement is weird.

Reply:
I'm curious... what would happen in the case of those editors and IDEs which automatically force the code on to a new line when the width of the line gets too long?

Can just imagine in those cases and the cases of the code shown above by @Irhn it will get very confusing.